lib/reference-resolver.js

```diff
@@ -0,0 +1,134 @@
+const { splitReference } = require('./utils');
+
+/** Class used by Client to resolve FHIR References. Do not use this class
+  * directly, use Client#resolve instead.
+  */
+class ReferenceResolver {
+  /**
+   * Create a Reference Resolver.
+   *
+   * @param {Client} client FHIR Client
+   */
+  constructor(client) {
+    this.client = client;
+  }
+
+  /**
+   * Resolve a reference and return FHIR resource
+   *
+   * From: http://hl7.org/fhir/STU3/references.html, a reference can be: 1)
+   * absolute URL, 2) relative URL or 3) an internal fragement.  In the case of
+   * (2), there are rules on resolving references that are in a FHIR bundle.
+   *
+   * @async
+   *
+   * @example
+   *
+   * // Always does a new http request
+   * client.resolve("http://test.com/fhir/Patient/1").then((patient) => {
+   *   console.log(patient);
+   * });
+   *
+   * // Always does a new http request, using the client.baseUrl
+   * client.resolve("Patient/1").then((patient) => {
+   *   console.log(patient);
+   * });
+   *
+   * // Try to resolve a patient in the bundle, otherwise build request
+   * // at client.baseUrl
+   * client.resolve("Patient/1", bundle).then((patient) => {
+   *   console.log(patient);
+   * });
+   *
+   * // Resolve a patient contained in someResource (see:
+   * // http://hl7.org/fhir/STU3/references.html#contained)
+   * client.resolve("#patient-1", someResource).then((patient) => {
+   *   console.log(patient);
+   * });
+   *
+   * @param {String} reference the FHIR reference
+   * @param {Object} [context] bundle or FHIR resource, optional
+   *
+   * @return {Promise<Object>} FHIR resource
+   */
+  async resolve(reference, context) {
+    if (context === undefined) {
+      if (reference.startsWith('http')) {
+        return this.resolveAbsoluteReference(reference);
+      }
+      return this.client.httpClient.get(reference);
+    }
+    if (reference.startsWith('#')) {
+      return this.resolveContainedReference(reference, context);
+    }
+    return this.resolveBundleReference(reference, context);
+  }
+
+  /**
+   * Private function for Client to resolve absolute references
+   *
+   * @private
+   *
+   * @async
+   *
+   * @param {String} reference the FHIR reference
+   *
+   * @return {Promise<Object>} FHIR resource
+   */
+  async resolveAbsoluteReference(reference) {
+    if (reference.startsWith(this.client.baseUrl)) {
+      return this.client.httpClient.get(reference);
+    }
+    const { baseUrl, resourceType, id } = splitReference(reference);
+    const { Client } = require('./client'); // eslint-disable-line global-require
+    return (new Client({ baseUrl })).read({ resourceType, id });
+  }
+
+  /**
+   * Private function for Client to resolve contained references
+   *
+   * @private
+   *
+   * @param {String} reference the FHIR reference
```

Comment:
I've been wonder if JSDocs should use `String` or `string` -- our existing code uses lowercase `string` and capitalized `Object`  

(which seems to match what the jsdoc github examples use?)

Reply:
https://github.com/jsdoc3/jsdoc/issues/1066 -- let's do this?

Reply:
And this: https://github.com/jsdoc3/jsdoc/issues/1046